o1heap/o1heap.c

```diff
@@ -401,6 +420,10 @@ void o1heapFree(O1HeapInstance* const handle, void* const pointer)
         O1HEAP_ASSERT(frag->header.size <= handle->diagnostics.capacity);
         O1HEAP_ASSERT((frag->header.size % FRAGMENT_SIZE_MIN) == 0U);
 
+#ifdef O1HEAP_TRACE
+        o1heapFreeTrace(handle, pointer, frag->header.size);
+#endif
```

Comment:
Preprocessor conditionals could be avoided if we replaced the functions with:

```c
#ifndef O1HEAP_TRACE_ALLOCATE
#   define O1HEAP_TRACE_ALLOCATE(handle, pointer, size)
#endif
#ifndef O1HEAP_TRACE_FREE
#   define O1HEAP_TRACE_FREE(handle, pointer, size)
#endif
```

Then the user would provide the required callbacks via `O1HEAP_CONFIG_HEADER`, similar to some other C libraries (LittleFS or Unity for example):

```c
// My custom config header: o1heap_config.h
// Add -DO1HEAP_CONFIG_HEADER=o1heap_config.h to the build flags

extern void o1heapAllocateTrace(O1HeapInstance* const handle, void* const allocated_memory, size_t size_bytes);
extern void o1heapFreeTrace(O1HeapInstance* const handle, void* const freed_memory, size_t size_bytes);

#define O1HEAP_TRACE_ALLOCATE(handle, pointer, size) \
    o1heapAllocateTrace(handle, pointer, trace)
#define O1HEAP_TRACE_FREE(handle, pointer, size) \
    o1heapFreeTrace(handle, pointer, size)
```

Why this is better:

- The code is not contaminated with conditional preprocessor blocks.
- Fewer API entities are introduced.